mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -1,149 +1,505 @@
-        watch_only: bool,
-        addresses: &[&str],
-    ) -> UtxoRpcFut<Vec<ZUnspent>>;
+    scan_cached_blocks(consensus_params, &db_guard.blocks_db, &mut db_data, None)?;
+    Ok(())
+}
 
-    fn z_send_many(&self, from_address: &str, send_to: Vec<ZSendManyItem>) -> UtxoRpcFut<String>;
+fn is_tx_imported(conn: &Connection, tx_id: TxId) -> bool {
+    const QUERY: &str = "SELECT id_tx FROM transactions WHERE txid = ?1;";
+    match query_single_row(conn, QUERY, [tx_id.0.to_vec()], |row| row.get::<_, i64>(0)) {
+        Ok(Some(_)) => true,
+        Ok(None) | Err(_) => false,
+    }
+}
 
-    fn z_import_key(&self, key: &str) -> UtxoRpcFut<()>;
+pub struct SaplingSyncRespawnGuard {
+    pub(super) sync_handle: Option<SaplingSyncLoopHandle>,
+    pub(super) abort_handle: Arc<Mutex<AbortOnDropHandle>>,
 }
 
-impl ZRpcOps for NativeClient {
-    fn z_get_balance(&self, address: &str, min_conf: u32) -> UtxoRpcFut<MmNumber> {
-        let fut = rpc_func!(self, "z_getbalance", address, min_conf);
-        Box::new(fut.map_to_mm_fut(UtxoRpcError::from))
+impl Drop for SaplingSyncRespawnGuard {
+    fn drop(&mut self) {
+        if let Some(handle) = self.sync_handle.take() {
+            *self.abort_handle.lock() = spawn_abortable(light_wallet_db_sync_loop(handle));
+        }
     }
+}
 
-    fn z_get_send_many_status(&self, op_ids: &[&str]) -> UtxoRpcFut<Vec<ZOperationStatus<ZOperationTxid>>> {
-        let fut = rpc_func!(self, "z_getoperationstatus", op_ids);
-        Box::new(fut.map_to_mm_fut(UtxoRpcError::from))
+impl SaplingSyncRespawnGuard {
+    pub(super) fn watch_for_tx(&mut self, tx_id: TxId) {
+        if let Some(ref mut handle) = self.sync_handle {
+            handle.watch_for_tx = Some(tx_id);
+        }
     }
 
-    fn z_list_unspent(
-        &self,
-        min_conf: u32,
-        max_conf: u32,
-        watch_only: bool,
-        addresses: &[&str],
-    ) -> UtxoRpcFut<Vec<ZUnspent>> {
-        let fut = rpc_func!(self, "z_listunspent", min_conf, max_conf, watch_only, addresses);
-        Box::new(fut.map_to_mm_fut(UtxoRpcError::from))
+    #[inline]
+    pub(super) fn current_block(&self) -> BlockHeight { self.sync_handle.as_ref().expect("always Some").current_block }
+}
+
+pub enum SyncStatus {
+    UpdatingBlocksCache {
+        current_scanned_block: u64,
+        latest_block: u64,
+    },
+    BuildingWalletDb,
+    Finished {
+        block_number: u64,
+    },
+}
+
+pub struct SaplingSyncLoopHandle {
+    pub(super) current_block: BlockHeight,
+    grpc_client: CompactTxStreamerClient<Channel>,
+    db: WalletDbShared,
+    consensus_params: ZcoinConsensusParams,
+    /// Notifies about sync status without stopping the loop, e.g. on coin activation
+    sync_status_notifier: AsyncSender<SyncStatus>,
+    /// If new tx is required to be generated, we stop the sync and respawn it after tx is sent
+    /// This watcher waits for such notification
+    on_tx_gen_watcher: AsyncReceiver<OneshotSender<Self>>,
```

Comment:
It's a bit complicated at the first glance, but seems to be the perfect pattern for the task!

Reply:
Thanks :)